service/mapper/transitions_test.go

```diff
@@ -0,0 +1,1233 @@
+		load := &mocks.Loader{}
+		chain := &mocks.Chain{}
+		feed := &mocks.Feeder{}
+		index := &mocks.Writer{}
+
+		tr := NewTransitions(testLog, load, chain, feed, index)
+
+		assert.NotNil(t, tr)
+		assert.Equal(t, load, tr.load)
+		assert.Equal(t, chain, tr.chain)
+		assert.Equal(t, feed, tr.feed)
+		assert.Equal(t, index, tr.index)
+		assert.NotNil(t, tr.once)
+		assert.Zero(t, tr.cfg)
+	})
+
+	t.Run("nominal case, with options", func(t *testing.T) {
+		load := &mocks.Loader{}
+		chain := &mocks.Chain{}
+		feed := &mocks.Feeder{}
+		index := &mocks.Writer{}
+
+		tr := NewTransitions(testLog, load, chain, feed, index,
+			WithIndexCommit(true),
+			WithIndexEvents(true),
+			WithIndexHeader(true),
+			WithIndexPayloads(true),
+			WithIndexTransactions(true),
+		)
+
+		assert.NotNil(t, tr)
+		assert.Equal(t, load, tr.load)
+		assert.Equal(t, chain, tr.chain)
+		assert.Equal(t, feed, tr.feed)
+		assert.Equal(t, index, tr.index)
+		assert.NotNil(t, tr.once)
+
+		assert.NotZero(t, tr.cfg)
+		assert.True(t, tr.cfg.IndexTransactions)
+		assert.True(t, tr.cfg.IndexEvents)
+		assert.True(t, tr.cfg.IndexHeader)
+		assert.True(t, tr.cfg.IndexPayloads)
+		assert.True(t, tr.cfg.IndexTransactions)
+	})
+}
+
+func TestTransitions_BootstrapState(t *testing.T) {
+	testHeight := uint64(42)
+
+	testTrie, err := trie.NewMTrie(testRoot)
+	require.NoError(t, err)
+
+	forest := &mocks.Forest{
+		SaveFunc: func(_ *trie.MTrie, _ []ledger.Path, _ flow.StateCommitment) {},
+	}
+	chain := &mocks.Chain{
+		RootFunc: func() (uint64, error) {
+			return testHeight, nil
+		},
+	}
+	load := &mocks.Loader{
+		CheckpointFunc: func() (*trie.MTrie, error) {
+			return testTrie, nil
+		},
+	}
+
+	tr := &Transitions{
+		load:  load,
+		chain: chain,
+	}
+
+	t.Run("nominal case", func(t *testing.T) {
+		t.Parallel()
+
+		// Copy state in local scope so that we can override its SaveFun without impacting other
+		// tests running in parallel.
+		var saveCalled bool
+		f := &mocks.Forest{
+			SaveFunc: func(tree *trie.MTrie, paths []ledger.Path, parent flow.StateCommitment) {
+				if !saveCalled {
+					assert.True(t, tree.IsEmpty())
+					assert.Nil(t, paths)
+					assert.Zero(t, parent)
+					saveCalled = true
+					return
+				}
+				assert.NotZero(t, tree)
+				assert.Len(t, paths, 3) // Expect the three paths from leaves.
+				assert.NotZero(t, parent)
+			},
+		}
+
+		st := &State{
+			forest: f,
+			status: StatusEmpty,
+		}
+
+		err = tr.BootstrapState(st)
+		assert.NoError(t, err)
+	})
+
+	t.Run("invalid state", func(t *testing.T) {
+		t.Parallel()
+
+		invalidState := &State{
+			status: StatusIndexed,
+		}
+
+		err = tr.BootstrapState(invalidState)
+		assert.Error(t, err)
+	})
+
+	t.Run("handles failure to get root height", func(t *testing.T) {
+		t.Parallel()
+
+		failingChain := &mocks.Chain{
+			RootFunc: func() (uint64, error) {
+				return 0, mocks.DummyError
+			},
+		}
+
+		failingTr := &Transitions{
+			chain: failingChain,
+		}
```

Comment:
This is the kind of case that we discussed. I think the `Transitions` component should always have _all_ the dependencies injected that are required for the `BootstrapState` transition here. Same for all other transitions.

Reply:
There could be a small helper to build the transactions with all dependencies in happy state here, and you just use that literally _everywhere_, and then you can just switch the component you want with a failing one. Should also reduce redundancy in code.

Reply:
I made a helper `baselineFSM(t *testing.T, status Status) (*Transition, *State) {}` for that. I think it's neater indeed. It makes it much clearer what exactly we need to change to make a test break 👍